Lib/ctypes/test/test_parameters.py

```diff
@@ -180,6 +181,25 @@ def test_abstract(self):
         self.assertRaises(TypeError, _Pointer.from_param, 42)
         self.assertRaises(TypeError, _SimpleCData.from_param, 42)
 
+    @test.support.cpython_only
+    def test_issue31311(self):
+        # __setstate__ should neither raise a SystemError nor crash in case
+        # of a bad __dict__.
+        from ctypes import Structure
+
+        class BadStruct(Structure):
+            def __dict__(self):
```

Comment:
I think it would be better to use a property. This doesn't work in any case, don't affect instance's dict, but at least looks more realistic.